### PR TITLE
[WIP] Refactor codegen module to allow different materailization

### PR DESCRIPTION
**Pull request by sklam**
Our Non-CPU backends rely on 3-rd party llvm tools that is older than the llvm version llvmlite link to.  This patches the codegen module to allow different materialization of llvm module.

### CODE REVIEW

**Comment by sklam** [comment]
This currently has the changes made in https://github.com/numba/numba/tree/hsa

File: numba/targets/codegen.py [hunk 1]
```diff
@@ -227,10 +229,13 @@ class BaseCPUCodegen(object):
     def __init__(self, module_name):
         self._libraries = set()
         self._data_layout = None
-        self._llvm_module = ll.parse_assembly(
-            str(self._create_empty_module(module_name)))
+        self._llvm_module = self._materialize_module(
+            self._create_empty_module(module_name))
         self._init(self._llvm_module)
 
+    def _materialize_module(self, ir_module):
```
Comment (gmarkall) [inline numba/targets/codegen.py]: Maybe just a nitpick, but this is used outside the `BaseCPUCodegen` class - should it be `materialize_module` (no underscore)?

Comment (sklam) [inline numba/targets/codegen.py]: Yes, you're right.

Comment (sklam) [inline numba/targets/codegen.py]: There are also `_create_empty_module` in https://github.com/numba/numba/pull/948/files#diff-ec37b531d0f5e6e4ff2bb91a63bd651dR35

**Comment by sklam** [comment]
At this point, this PR is to let everyone be aware of the need in the HSA backend and it is not final.  I will probably make another round of refactor with consideration to the needs of the CUDA backend as well.

**Pull request closed without merging**